### PR TITLE
Tweak: Make Elementor compatible to WebP [ED-7895]

### PR DESCRIPTION
**Pull request by maork-elementor**
via Performance Lab plugin

### CODE REVIEW

File: includes/utils.php [hunk 1]
```diff
@@ -790,4 +790,20 @@ public static function file_get_contents( $file, ...$args ) {
 
 		return file_get_contents( $file, ...$args );
 	}
+
+	public static function performace_lab_get_webp_src( $attachment_id, $size = 'full' ) {
+		if ( function_exists( 'webp_uploads_img_tag_update_mime_type' ) ) {
+						$webp_option_name = 'site-health/webp-support';
+						$perflab_modules_settings = get_option( 'perflab_modules_settings' );
```
Comment (matipojo) [inline includes/utils.php:797]: PHPCS

File: includes/utils.php [hunk 2]
```diff
@@ -790,4 +790,20 @@ public static function file_get_contents( $file, ...$args ) {
 
 		return file_get_contents( $file, ...$args );
 	}
+
+	public static function performace_lab_get_webp_src( $attachment_id, $size = 'full' ) {
```
Comment (matipojo) [inline includes/utils.php:794]: Please cover with tests

**Review by matipojo — CHANGES_REQUESTED** [review]

File: includes/utils.php (hunk 1 above)
Comment (maork-elementor) [reply includes/utils.php]: done

File: includes/utils.php (hunk 2 above)
Comment (maork-elementor) [reply includes/utils.php]: done

File: core/modules-manager.php [hunk 3]
```diff
@@ -98,6 +98,7 @@ public function get_modules_names() {
 			'web-cli',
 			'promotions',
 			'notes',
+			'third-party-customization',
```
Comment (matipojo) [inline core/modules-manager.php:101]:
```suggestion
			'performace-lab',
```

**Review by matipojo — CHANGES_REQUESTED** [review]

File: core/modules-manager.php (hunk 3 above)
Comment (maork-elementor) [reply core/modules-manager.php:101]: done

File: modules/third-party-customization/module.php [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+use Elementor\Core\Base\Module as BaseModule;
+use Elementor\Plugin;
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly
+}
+
+class Module extends BaseModule {
+
+
```
[the remark this answers is not in the document]
Comment (maork-elementor) [reply modules/third-party-customization/module.php:13]: done

File: modules/third-party-customization/module.php [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+	public function __construct() {
+		parent::__construct();
+		add_filter('elementor/css-file/css_property', function( $value, $css_property, $matches, $control ) {
+			if ( 0 === strpos( $css_property, 'background-image' ) && '{{URL}}' === $matches[0] ) {
+				$value['url'] = $this->performace_lab_get_webp_src( $value['id'], 'full', $value['url'] );
+			}
+			return $value;
+		}, 10, 4 );
```
[the remark this answers is not in the document]
Comment (maork-elementor) [reply modules/third-party-customization/module.php:42]: done

File: core/files/css/base.php [hunk 6]
```diff
@@ -343,6 +343,8 @@ public function add_control_rules( array $control, array $controls_stack, callab
 
 						$parsed_value = '';
 
+						$value = apply_filters( 'elementor/css-file/css_property', $value, $css_property, $matches , $control );
```
[the remark this answers is not in the document]
Comment (maork-elementor) [reply core/files/css/base.php]: done

File: modules/performance-lab/module.php [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+use Elementor\Core\Base\Module as BaseModule;
+use Elementor\Plugin;
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly
+}
+
+class Module extends BaseModule {
+
+
```
Comment (matipojo) [inline modules/performance-lab/module.php:13]:
```suggestion

```

File: modules/performance-lab/module.php [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+	public function get_name() {
+		return 'performance-lab';
+	}
+
+	private function does_performance_lab_is_active() {
+		if ( function_exists( 'webp_uploads_img_tag_update_mime_type' ) ) {
+			$webp_option_name = 'site-health/webp-support';
+			$perflab_modules_settings = get_option( 'perflab_modules_settings' );
+			if ( isset( $perflab_modules_settings ) && isset( $perflab_modules_settings[ $webp_option_name ] ) &&
+							'1' === $perflab_modules_settings[ $webp_option_name ]['enabled'] ) {
```
Comment (matipojo) [inline modules/performance-lab/module.php:23]: Move all strings to constants

File: modules/performance-lab/module.php [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+class Module extends BaseModule {
+
+
+	public function get_name() {
+		return 'performance-lab';
+	}
+
+	private function does_performance_lab_is_active() {
```
Comment (matipojo) [inline modules/performance-lab/module.php:18]:
```suggestion
	private function is_performance_lab_is_active() {
```

File: modules/performance-lab/module.php [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+			if ( isset( $perflab_modules_settings ) && isset( $perflab_modules_settings[ $webp_option_name ] ) &&
+							'1' === $perflab_modules_settings[ $webp_option_name ]['enabled'] ) {
+				return true;
+			}
+		}
+		return false;
+	}
+
+	private function performance_lab_get_webp_src( $attachment_id, $size = 'full', $url ) {
+		if ( $this->does_performance_lab_is_active() ) {
```
Comment (matipojo) [inline modules/performance-lab/module.php:31]: Remove, the filter is set only when it's active
```suggestion
```

File: tests/phpunit/elementor/modules/performance-lab/test-module.php [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+	public function test_performance_lab_get_webp_src() {
+
+		$image_id = $this->factory()->attachment->create_upload_object( __DIR__ . '/assets/pixel.jpg' );
+		$image_url = wp_get_attachment_url( $image_id );
+		$reflection = new \ReflectionClass( PerformanceLab::class );
+		$method = $reflection->getMethod( 'performance_lab_get_webp_src' );
+		$method->setAccessible( true );
+		$performance_lab = new PerformanceLab();
+		$webp_src = $method->invokeArgs( $performance_lab, [ $image_id, 'full', $image_url ] );
+		$this->assertEquals( $image_url, $webp_src );
```
Comment (matipojo) [inline tests/phpunit/elementor/modules/performance-lab/test-module.php:18]: Please assert with hard-coded value (e.g. includes `/assets/pixel.webp`)

**Review by matipojo — CHANGES_REQUESTED** [review]

File: modules/performance-lab/module.php (hunk 7 above)
Comment (maork-elementor) [reply modules/performance-lab/module.php]: done

File: modules/performance-lab/module.php (hunk 8 above)
Comment (maork-elementor) [reply modules/performance-lab/module.php]: done

File: modules/performance-lab/module.php (hunk 9 above)
Comment (maork-elementor) [reply modules/performance-lab/module.php]: done

File: modules/performance-lab/module.php (hunk 10 above)
Comment (maork-elementor) [reply modules/performance-lab/module.php]: done

File: tests/phpunit/elementor/modules/performance-lab/test-module.php (hunk 11 above)
Comment (maork-elementor) [reply tests/phpunit/elementor/modules/performance-lab/test-module.php:18]: I don't understand what you're asking for.
Would you mind elaborating?

File: tests/phpunit/elementor/modules/performance-lab/test-module.php [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+
+	public function test_performance_lab_get_webp_src() {
+		$image_id = $this->factory()->attachment->create_upload_object( __DIR__ . '/assets/pixel.jpg' );
+		$image_url = wp_get_attachment_url( $image_id );
+		$reflection = new \ReflectionClass( PerformanceLab::class );
+		$method = $reflection->getMethod( 'performance_lab_get_webp_src' );
+		$method->setAccessible( true );
+		$performance_lab = new PerformanceLab();
+		$webp_src = $method->invokeArgs( $performance_lab, [ $image_id, 'full', $image_url ] );
+		$this->assertNotEquals( '.jpg', substr( $webp_src, -4 ) );
```
Comment (matipojo) [inline tests/phpunit/elementor/modules/performance-lab/test-module.php:17]: Assert it's `webp`

File: tests/phpunit/elementor/modules/performance-lab/test-module.php [hunk 13]
```diff
@@ -0,0 +1,19 @@
+<?php
+namespace Elementor\Testing\Modules\PerformanceLab;
+
+use ElementorEditorTesting\Elementor_Test_Base;
+use Elementor\Modules\PerformanceLab\Module as PerformanceLab;
+require_once( __DIR__ . '/mock.php' );
+class Elementor_Test_PerformanceLab extends Elementor_Test_Base {
+
+	public function test_performance_lab_get_webp_src() {
+		$image_id = $this->factory()->attachment->create_upload_object( __DIR__ . '/assets/pixel.jpg' );
```
Comment (matipojo) [inline tests/phpunit/elementor/modules/performance-lab/test-module.php:10]: Move to the assets folder

File: tests/phpunit/elementor/modules/performance-lab/mock.php [hunk 14]
```diff
@@ -0,0 +1,8 @@
+<?php
+
+function webp_uploads_img_tag_update_mime_type( $image_object, $size, $attachment_id ) {
+    // Image_object & size are not used in this function (mocked)
+	$image_url = wp_get_attachment_url( $attachment_id );
+    $image_url = preg_replace( '/\.[^.]+$/', '.webp', $image_url );
```
Comment (matipojo) [inline tests/phpunit/elementor/modules/performance-lab/mock.php:6]: PHPCS

File: modules/performance-lab/module.php [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+		$image_src = call_user_func( self::PERFORMANCE_LAB_FUNCTION_NAME, $image_object[0], 'webp', $attachment_id );
+		if ( ! empty( $image_src ) ) {
+			return $image_src;
+		}
+		return $url;
+	}
+
+
+	public function __construct() {
+
```
Comment (matipojo) [inline modules/performance-lab/module.php:42]: PHPCS

File: modules/performance-lab/module.php [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+		parent::__construct();
+
+		if ( $this->is_performance_lab_is_active() ) {
+			add_filter('elementor/files/css/property', function( $value, $css_property, $matches, $control ) {
+				if ( 0 === strpos( $css_property, 'background-image' ) && '{{URL}}' === $matches[0] ) {
+					$value['url'] = $this->performance_lab_get_webp_src( $value['id'], 'full', $value['url'] );
+				}
+				return $value;
```
Comment (matipojo) [inline modules/performance-lab/module.php:50]: Extract to a private method

**Review by matipojo — CHANGES_REQUESTED** [review]

File: tests/phpunit/elementor/modules/performance-lab/test-module.php (hunk 11 above)
Comment (maork-elementor) [reply tests/phpunit/elementor/modules/performance-lab/test-module.php]: done

File: tests/phpunit/elementor/modules/performance-lab/test-module.php (hunk 13 above)
Comment (maork-elementor) [reply tests/phpunit/elementor/modules/performance-lab/test-module.php:10]: done

File: tests/phpunit/elementor/modules/performance-lab/mock.php (hunk 14 above)
Comment (maork-elementor) [reply tests/phpunit/elementor/modules/performance-lab/mock.php:6]: done

File: modules/performance-lab/module.php (hunk 15 above)
Comment (maork-elementor) [reply modules/performance-lab/module.php:42]: done

File: modules/performance-lab/module.php (hunk 16 above)
Comment (maork-elementor) [reply modules/performance-lab/module.php:50]: done

File: modules/performance-lab/module.php (hunk 15 above)
Comment (matipojo) [reply modules/performance-lab/module.php:50]: Where?

File: tests/phpunit/elementor/modules/performance-lab/test-module.php [hunk 17]
```diff
@@ -0,0 +1,20 @@
+<?php
+namespace Elementor\Testing\Modules\PerformanceLab;
+
+use ElementorEditorTesting\Elementor_Test_Base;
+use Elementor\Modules\PerformanceLab\Module as PerformanceLab;
+require_once __DIR__ . '/mock.php';
+class Elementor_Test_PerformanceLab extends Elementor_Test_Base {
+
+	public function test_performance_lab_get_webp_src() {
+		$root_path = dirname( dirname( dirname( dirname( dirname( __DIR__ ) ) ) ) );
```
Comment (matipojo) [inline tests/phpunit/elementor/modules/performance-lab/test-module.php:10]: Use ELEMENTOR_PATH

File: modules/performance-lab/module.php [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+	private function performance_lab_get_webp_src( $attachment_id, $size = 'full', $url ) {
+		$image_object = wp_get_attachment_image_src( $attachment_id, $size );
+		$image_src = call_user_func( self::PERFORMANCE_LAB_FUNCTION_NAME, $image_object[0], 'webp', $attachment_id );
+		if ( ! empty( $image_src ) ) {
+			return $image_src;
+		}
+		return $url;
+	}
+
+	private function apply_css_filter() {
```
Comment (matipojo) [inline modules/performance-lab/module.php:40]:
```suggestion
	private function add_css_filter() {
```

**Review by matipojo — CHANGES_REQUESTED** [review]

File: tests/phpunit/elementor/modules/performance-lab/test-module.php (hunk 17 above)
Comment (maork-elementor) [reply tests/phpunit/elementor/modules/performance-lab/test-module.php:10]: done

File: modules/performance-lab/module.php (hunk 18 above)
Comment (maork-elementor) [reply modules/performance-lab/module.php:40]: done

File: modules/performance-lab/module.php (hunk 15 above)
Comment (maork-elementor) [reply modules/performance-lab/module.php:50]: I activated the PHPCS, and it looks fine, no error or warnings

File: tests/phpunit/elementor/modules/performance-lab/mock.php [hunk 19]
```diff
@@ -0,0 +1,8 @@
+<?php
+
+function webp_uploads_img_tag_update_mime_type( $image_object, $size, $attachment_id ) {
```
Comment (matipojo) [inline tests/phpunit/elementor/modules/performance-lab/mock.php:3]: Where do you use it?

**Review by matipojo — CHANGES_REQUESTED** [review]

File: modules/performance-lab/module.php [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+		add_filter('elementor/files/css/property', function( $value, $css_property, $matches, $control ) {
+			if ( 0 === strpos( $css_property, 'background-image' ) && '{{URL}}' === $matches[0] ) {
+				$value['url'] = $this->performance_lab_get_webp_src( $value['id'], 'full', $value['url'] );
+			}
+			return $value;
+		}, 10, 4 );
+	}
+
+	public function __construct() {
+
```
Comment (matipojo) [inline modules/performance-lab/module.php:50]:
```suggestion
```

File: modules/performance-lab/module.php [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+				Plugin::$instance->files_manager->clear_cache();
+			}
+		} );
+
+		add_action( 'deactivated_plugin', function( $plugin ) {
+			if ( 'performance-lab/load.php' === $plugin ) {
+				Plugin::$instance->files_manager->clear_cache();
+			}
+		} );
+
```
Comment (matipojo) [inline modules/performance-lab/module.php:67]:
```suggestion
```

File: modules/performance-lab/module.php [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+		add_action( 'deactivated_plugin', function( $plugin ) {
+			if ( 'performance-lab/load.php' === $plugin ) {
+				Plugin::$instance->files_manager->clear_cache();
+			}
+		} );
```
Comment (matipojo) [inline modules/performance-lab/module.php:66]:
```suggestion
	  if ( is_admin() ) {
		  add_action( 'activated_plugin', function( $plugin ) {
			  if ( 'performance-lab/load.php' === $plugin ) {
				  Plugin::$instance->files_manager->clear_cache();
			  }
		  } );
  
		  add_action( 'deactivated_plugin', function( $plugin ) {
			  if ( 'performance-lab/load.php' === $plugin ) {
				  Plugin::$instance->files_manager->clear_cache();
			  }
		  } );
	}
```

File: modules/performance-lab/module.php [hunk 23]
```diff
@@ -0,0 +1,70 @@
...
+				  Plugin::$instance->files_manager->clear_cache();
+			  }
+		  } );
+  
+		  add_action( 'deactivated_plugin', function( $plugin ) {
+			  if ( 'performance-lab/load.php' === $plugin ) {
+				  Plugin::$instance->files_manager->clear_cache();
+			  }
+		  } );
+	}
```
Comment (matipojo) [inline modules/performance-lab/module.php:68]:
```suggestion
		}
```

File: modules/performance-lab/module.php [hunk 24]
```diff
@@ -0,0 +1,70 @@
...
+				  Plugin::$instance->files_manager->clear_cache();
+			  }
+		  } );
+  
+		  add_action( 'deactivated_plugin', function( $plugin ) {
+			  if ( 'performance-lab/load.php' === $plugin ) {
+				  Plugin::$instance->files_manager->clear_cache();
+			  }
+		  } );
+		}
```
Comment (matipojo) [inline modules/performance-lab/module.php:68]:
```suggestion
	  i}
```

File: modules/performance-lab/module.php [hunk 25]
```diff
@@ -0,0 +1,70 @@
...
+			$value['url'] = $this->performance_lab_get_webp_src( $value['id'], 'full', $value['url'] );
+		}
+		return $value;
+	}
+
+	public function __construct() {
+		parent::__construct();
+
+		if ( $this->is_performance_lab_is_active() ) {
+			add_filter('elementor/files/css/property', function( $value, $css_property, $matches ) {
```
Comment (matipojo) [inline modules/performance-lab/module.php:51]:
```suggestion
			add_filter( 'elementor/files/css/property', function( $value, $css_property, $matches ) {
```

**Review by matipojo — APPROVED** [review]

**Comment by adamsilverstein** [comment]
Hi @matipojo and @maork-elementor 👋🏼 ,

Thanks for working on adding direct support for the Performance Lab plugin in Elementor - its fantastic seeing the community embrace performance as a core initiative. Kudos to your team for supporting the effort!

I wanted to point out that since https://github.com/WordPress/performance/pull/527 the Performance Lab plugin only generates WebP images by default. So (unless that default is changed), the replacements you introduced in this PR won't strictly be necessary any longer - since only WebP images are generated, they will be used by default everywhere and although multi-mime is supported in the plugin, it isn't used by default.

Have you considered adding the capability of enabling WebP output directly in Elementor's settings? This would give all Elementor users (with a supporting server) access to WebP and can be enabled with a small bit of code since WP 5.8 (see example [gist](https://gist.github.com/adamsilverstein/89c2d0c2128422c93e4fa6d8d0489be0)). Curious what you think!

**Pull request merged**